### PR TITLE
Implement #1456: ansible dynamic inventory for tsh ls output

### PR DESCRIPTION
**Pull request by halfa**
This pull request is a rough implementation of the feature described in #1456. It allows the `tsh` client to output the node list in a format compatible with `ansible` dynamic inventory as described on [the ansible doc](http://docs.ansible.com/ansible/latest/dev_guide/developing_inventory.html). It also implements the `--list` and `--host` flags.

This is how you can test it:

Create a script with your `tsh` configuration (here `inventory.sh`):
```
#!/bin/bash
tsh --proxy=<url>:<port> --user=<user> --login=<login> ls --format=json "$@"
```
Then we can pass this file as a dynamic inventory provider to `ansible`. Don't forget to `tsh login` (or `export $(tsh agent)` if you use the gnome keyring) beforehand!
```
ansible -i inventory.sh -m ping -e ansible_ssh_user=<user> -e ansible_ssh_port=<teleport_node_port> all
```

There is also a second flag, `format=ansible-static`, which output a static inventory if necessary as described in #1456, and some tests for JSON marshaling stability.

As of why I choose to implement groups based on labels and not to populate `_meta`, that's because it was easier for writing playbooks, but I'll be happy to change this behavior if needed. I also need to rebase this against master, but as I'm still using `2.3` and this code in production I though it should be easier to review here. That's also the reason why the log is a bit verbose.

(As @klizhentas said back in november, I took my time for this one. It actually slept almost finished for a month 😄)

### CODE REVIEW

File: lib/ansible/inventory.go [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+	// this gonna become "_meta": { "hostvars": { "host": {"var": value}}}
+	// so the 2 top level (Meta and Hotvars) have only one key to match the struct
+	// yes, the type is stupid, but blame python devs not me
+	type Meta map[string]map[string]map[string]string
+	meta := make(Meta)
+	meta["hostvars"] = make(map[string]map[string]string)
+	inventory["_meta"] = meta
+	out, err := json.Marshal(inventory)
+	if err != nil {
+		return "", fmt.Errorf("cannot encode JSON objet: %s", err)
```
Comment (klizhentas) [inline lib/ansible/inventory.go]: return "", trace.Wrap(err, "can not encode JSON object")

**Comment by kontsevoy** [comment]
@halfa this is interesting. we cannot accept this PR as-is though because you are pushing it into branch/2.3 (it's a frozen branch).

Instead, can you please submit another PR which:

1. Is based on teleport master branch.
2. Targets `tsh-ansible` branch in Teleport repo (which I just created for you).

We'll make some changes to flag naming. I definitely am not a fan of calling Ansible-specific output "json", but otherwise everything looks sensible, thank you so much for implementing this.

**Review by klizhentas — COMMENTED** [review]
here is the first round of comments, will do a second pass this week

**Comment by stanchan** [comment]
I've been testing this on branch/2.4 and it works great! Thanks for implementing this @halfa. Just wanted to suggest a change in how the label works. It doesn't seem practical to create groups for every label in teleport. This is especially true if you have commands associated with some of them. Maybe having any label with Ansible as the value be assigned to that given group or a comma delimited set of groups for each node be a more useful feature?

File: lib/ansible/inventory.go [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+	"github.com/gravitational/teleport/lib/services"
+)
+
+// Inventory matches the JSON struct needed for DynamicInventoryList
+type Inventory map[string]interface{}
+
+// Group gather hosts and variables common to them
+type Group struct {
+	Hosts []string          `json:"hosts"`
```
[the remark this answers is not in the document]
Comment (halfa) [reply lib/ansible/inventory.go]: Fixed

File: lib/ansible/inventory.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+//         "vars": {
+//             "a": true
+//         }
+//     },
+//     "_meta": {
+//         "hostvars": {}
+//     }
+// }
+// ```
+func DynamicInventoryList(nodes []services.Server) (string, error) {
```
[the remark this answers is not in the document]
Comment (halfa) [reply lib/ansible/inventory.go]: I agree

File: lib/ansible/inventory.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+			Hosts: hosts,
+			Vars:  make(map[string]string),
+		}
+	}
+
+	// Meta is a special group with information on each host
+	// this gonna become "_meta": { "hostvars": { "host": {"var": value}}}
+	// so the 2 top level (Meta and Hotvars) have only one key to match the struct
+	// yes, the type is stupid, but blame python devs not me
+	type Meta map[string]map[string]map[string]string
```
[the remark this answers is not in the document]
Comment (halfa) [reply lib/ansible/inventory.go]: Yep, I think I planned to use the type elsewhere but it wasn't needed in the end. Fixed.

File: lib/ansible/inventory.go [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+	meta := make(Meta)
+	meta["hostvars"] = make(map[string]map[string]string)
+	inventory["_meta"] = meta
+	out, err := json.Marshal(inventory)
+	if err != nil {
+		return "", fmt.Errorf("cannot encode JSON objet: %s", err)
+	}
+	return string(out) + "\n", nil
```
[the remark this answers is not in the document]
Comment (halfa) [reply lib/ansible/inventory.go]: True, fixed

File: lib/ansible/inventory.go [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+}
+
+// DynamicInventoryHost returns a JSON-formated ouput compatible with Ansible --host <string> flag
+//
+// (From ansible ref. doc)
+// When called with the arguments --host <hostname>, the script must print either an empty JSON hash/dictionary,
+// or a hash/dictionary of variables to make available to templates and playbooks.
+func DynamicInventoryHost(nodes []services.Server, host string) {
```
[the remark this answers is not in the document]
Comment (halfa) [reply lib/ansible/inventory.go]: Fixed

**Comment by halfa** [comment]
@klizhentas thanks for the feedback, your comments should be fixed with the last set of commit!
As @kontsevoy pointed, it's meaningless to work further on a frozen branch, I've opened a new PR against the `teleport/tsh-ansible` branch: #1608.

**Pull request closed without merging**